File: app/src/main/java/com/alaskalany/careershowcase/ui/contact/ContactListFragment.java
package com.alaskalany.careershowcase.ui.contact;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.alaskalany.careershowcase.R;
import com.alaskalany.careershowcase.databinding.FragmentContactListBinding;
import com.alaskalany.careershowcase.ui.ScrollToTop;
import com.alaskalany.careershowcase.viewmodel.ContactListViewModel;

/**
 * A fragment representing a list of Items.
 * <p/>
 */
public class ContactListFragment
        extends Fragment
        implements ScrollToTop {

    /**
     *
     */
    protected static final String ARG_COLUMN_COUNT = "column-count";

    /**
     *
     */
    private final ContactOnClickCallback mCallBack =
            item -> Toast.makeText(getContext(), "Clicked on ContactEntity Item", Toast.LENGTH_SHORT).show();

    /**
     *
     */
    protected FragmentContactListBinding mBinding;

    /**
     *
     */
    protected ContactAdapter mAdapter;

    /**
     *
     */
    protected int mColumnCount = 1;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public ContactListFragment() {

    }

    /**
     * @param columnCount
     *
     * @return
     */
    @SuppressWarnings("unused")
    public static ContactListFragment newInstance(int columnCount) {

        ContactListFragment fragment = new ContactListFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_COLUMN_COUNT, columnCount);
        fragment.setArguments(args);
        return fragment;
    }

    /**
     * @param inflater
     * @param container
     * @param savedInstanceState
     *
     * @return
     */
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_contact_list, container, false);
        setAdapter(new ContactAdapter(mCallBack));
        Context context = mBinding.getRoot().getContext();
        if (getColumnCount() <= 1) {
            mBinding.listContact.setLayoutManager(new LinearLayoutManager(context));
        } else {
            mBinding.listContact.setLayoutManager(new GridLayoutManager(context, getColumnCount()));
        }
        mBinding.listContact.setAdapter(getAdapter());
        return mBinding.getRoot();
    }

    /**
     * Called when the fragment's activity has been created and this
     * fragment's view hierarchy instantiated.  It can be used to do final
     * initialization once these pieces are in place, such as retrieving
     * views or restoring state.  It is also useful for fragments that use
     * {@link #setRetainInstance(boolean)} to retain their instance,
     * as this callback tells the fragment when it is fully associated with
     * the new activity instance.  This is called after {@link #onCreateView}
     * and before {@link #onViewStateRestored(Bundle)}.
     *
     * @param savedInstanceState If the fragment is being re-created from
     *                           a previous saved state, this is the state.
     */
    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {

        super.onActivityCreated(savedInstanceState);
        final ContactListViewModel _model = ViewModelProviders.of(this).get(ContactListViewModel.class);
        mBinding.setContactListViewModel(_model);
        _model.getContacts().observe(this, contactEntities -> {
            if (contactEntities != null) {
                mAdapter.setContactList(contactEntities);
            } else {
            }
            mBinding.executePendingBindings();
        });
    }

    /**
     * @return
     */
    protected int getColumnCount() {

        return mColumnCount;
    }

    /**
     * @param mColumnCount
     */
    protected void setColumnCount(int mColumnCount) {

        this.mColumnCount = mColumnCount;
    }

    /**
     * @return
     */
    protected ContactAdapter getAdapter() {

        return mAdapter;
    }

    /**
     * @param adapter
     */
    protected void setAdapter(ContactAdapter adapter) {

        this.mAdapter = adapter;
    }

    /**
     * @param context
     */
    @Override
    public void onAttach(Context context) {

        super.onAttach(context);
    }

    /**
     * @param savedInstanceState
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            setColumnCount(getArguments().getInt(ARG_COLUMN_COUNT));
        }
    }

    /**
     *
     */
    @Override
    public void onDetach() {

        super.onDetach();
    }

    @Override
    public void top() {

        mBinding.listContact.smoothScrollToPosition(0);
    }
}

File: app/src/main/java/com/alaskalany/careershowcase/ui/contact/ContactListFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2018 Ahmed AlAskalany
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.alaskalany.careershowcase.ui.contact

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.alaskalany.careershowcase.R
import com.alaskalany.careershowcase.databinding.FragmentContactListBinding
import com.alaskalany.careershowcase.entity.ContactEntity
import com.alaskalany.careershowcase.model.Contact
import com.alaskalany.careershowcase.ui.ScrollToTop
import com.alaskalany.careershowcase.viewmodel.ContactListViewModel

/**
 * A fragment representing a list of Items.
 *
 *
 */
/**
 * Mandatory empty constructor for the fragment manager to instantiate the
 * fragment (e.g. upon screen orientation changes).
 */
class ContactListFragment : Fragment(), ScrollToTop {

    /**
     *
     */
    private val contactOnClickCallback = ContactOnClickCallback {
        Toast.makeText(this@ContactListFragment.context, "Clicked on ContactEntity Item", Toast.LENGTH_SHORT)
                .show()
    }

    /**
     *
     */
    private var binding: FragmentContactListBinding? = null

    /**
     *
     */
    /**
     * @return
     */
    /**
     * @param adapter
     */
    protected var adapter: ContactAdapter? = null

    /**
     *
     */
    /**
     * @return
     */
    /**
     * @param mColumnCount
     */
    protected var columnCount = 1

    /**
     * @param context
     */
    override fun onAttach(context: Context?) {

        super.onAttach(context)
    }

    /**
     * @param savedInstanceState
     */
    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        if (arguments != null) {
            columnCount = arguments!!.getInt(ARG_COLUMN_COUNT)
        }
    }

    /**
     * @param inflater
     * @param container
     * @param savedInstanceState
     *
     * @return
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_contact_list, container, false)
        adapter = ContactAdapter(contactOnClickCallback)
        val context = binding!!.root
                .context
        if (columnCount <= 1) {
            binding!!.listContact.layoutManager = LinearLayoutManager(context)
        } else {
            binding!!.listContact.layoutManager = GridLayoutManager(context, columnCount)
        }
        binding!!.listContact.adapter = adapter
        return binding!!.root
    }

    /**
     * Called when the fragment's activity has been created and this
     * fragment's view hierarchy instantiated.  It can be used to do final
     * initialization once these pieces are in place, such as retrieving
     * views or restoring state.  It is also useful for fragments that use
     * [.setRetainInstance] to retain their instance,
     * as this skillOnClickCallback tells the fragment when it is fully associated with
     * the new activity instance.  This is called after [.onCreateView]
     * and before [.onViewStateRestored].
     *
     * @param savedInstanceState If the fragment is being re-created from
     * a previous saved state, this is the state.
     */
    override fun onActivityCreated(savedInstanceState: Bundle?) {

        super.onActivityCreated(savedInstanceState)
        val _model = ViewModelProviders.of(this)
                .get(ContactListViewModel::class.java)
        binding!!.contactListViewModel = _model
        _model.contacts
                .observe(this, Observer { contactEntities ->
                    if (contactEntities != null) {
                        adapter!!.setContactList(contactEntities)
                    } else {
                    }
                    binding!!.executePendingBindings()
                })
    }

    /**
     *
     */
    override fun onDetach() {

        super.onDetach()
    }

    override fun top() {

        binding!!.listContact.smoothScrollToPosition(0)
    }

    companion object {

        /**
         *
         */
        protected val ARG_COLUMN_COUNT = "column-count"

        /**
         * @param columnCount
         *
         * @return
         */
        fun newInstance(columnCount: Int): ContactListFragment {

            val fragment = ContactListFragment()
            val args = Bundle()
            args.putInt(ARG_COLUMN_COUNT, columnCount)
            fragment.arguments = args
            return fragment
        }
    }
}
